### PR TITLE
Message retraction

### PR DESCRIPTION
**Pull request by chrzaszcz**
Message retraction - implements [XEP-0424](https://xmpp.org/extensions/xep-0424.html#example-4)

- The functionality has been reviewed in the previous PR's: #2722, #2726 (see below)
- Documentation added, manual tests performed for the migrations for all three RDBMS

Note: migrations were tested related to 1ba383aad0ac as we are missing the description for previous DB script changes since 3.6.0.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[8234.1](https://travis-ci.org/esl/MongooseIM/jobs/686109789) / Erlang 22.0 / small_tests / 4f9b874c81ec466fb3a51fc15d769011efc993bf
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2739/8234/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2739/8234/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-05-12_13.37.52/lib.mongooseim.logs/run.2020-05-12_13.37.52/suite.log.html)

**Comment by mongoose-im** [comment]
[8235.1](https://travis-ci.org/esl/MongooseIM/jobs/686112285) / Erlang 22.0 / small_tests / 56e1b947adba3364a621004db161692fe579e78d
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2739/8235/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2739/8235/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-05-12_14.09.47/lib.mongooseim.logs/run.2020-05-12_14.09.47/suite.log.html)

File: doc/migrations/3.6.0_3.6.x.md [hunk 1]
```diff
@@ -45,3 +45,39 @@ If you want to use the new password format with a full set of supported SHA hash
...
+ALTER TABLE mam_message ADD COLUMN origin_id varchar;
+CREATE INDEX i_mam_message_username_jid_origin_id ON mam_message USING BTREE (user_id, remote_bare_jid, origin_id);
+
+ALTER TABLE mam_muc_message ADD COLUMN origin_id varchar;
+CREATE INDEX i_mam_muc_message_room_id_sender_id_origin_id ON mam_muc_message USING BTREE (room_id, sender_id, origin_id);
+```
+
+**MSSQL**
+
+Note: `i_mam_message_username_jid_id` was missing from the schema, this was fixed.
```
Comment (goddammit) [inline doc/migrations/3.6.0_3.6.x.md:73]:
```suggestion
Note: `i_mam_message_username_jid_id` was missing from the schema, this is now fixed.
```

File: doc/migrations/3.6.0_3.6.x.md [hunk 2]
```diff
@@ -45,3 +45,39 @@ If you want to use the new password format with a full set of supported SHA hash
 Otherwise, please make sure that you provide the right `sasl_mechanism` configuration, where the mechanism you authenticate with is compatible with the password format you store.
 
 For more details related to the new password format, please refer to [authentication](../../Advanced-configuration#authentication) and [SCRAM serialization](../../scram-serialization.md) sections.
+
+## Message retraction
+
+Support for [XEP-0424: Message Retraction](http://xmpp.org/extensions/xep-0424.html) requires a new column for the `origin_id` attribute of MAM messages, which allows MAM to identify the messages to retract. Indexes for this column are required for efficient queries. Only the messages stored after this change can be retracted.
```
Comment (michalwski) [inline doc/migrations/3.6.0_3.6.x.md:51]: Reading this sentence I can interpret it in a way that if I don't want the support for message retraction I don't have to add the column to the database. I think this assumption is not true, and we need the new column in order to archive the messages correctly regardless of the retraction functionality. Could you rephrase it in a way that this new column is mandatory?

File: priv/mssql2012.sql [hunk 3]
```diff
@@ -66,14 +66,21 @@ CREATE TABLE [dbo].[mam_message](
...
+	[origin_id] [nvarchar](250) NULL,
  CONSTRAINT [PK_mam_message_user_id] PRIMARY KEY CLUSTERED
 (
 	[user_id] ASC,
 	[id] ASC
 )WITH (PAD_INDEX = OFF, STATISTICS_NORECOMPUTE = OFF, IGNORE_DUP_KEY = OFF, ALLOW_ROW_LOCKS = ON, ALLOW_PAGE_LOCKS = ON) ON [PRIMARY]
 ) ON [PRIMARY] TEXTIMAGE_ON [PRIMARY]
+GO
+
+CREATE INDEX i_mam_message_username_jid_id ON mam_message (user_id, remote_bare_jid, id);
```
Comment (michalwski) [inline priv/mssql2012.sql:78]: Good catch!

File: doc/migrations/3.6.0_3.6.x.md (hunk 2 above)
Comment (chrzaszcz) [reply doc/migrations/3.6.0_3.6.x.md]: > A corner case just came to my mind, which as far as I can see we are handling in a good enough way.
> A user sends two different messages with the same `origin-id`, according to current schema they will be stored. When retracting, only the newer message will be retracted.
> @chrzaszcz do you think it'd be worth to document this behaviour and advise the client to use ids as unique as possible, like `UUIDs`?

Retracting only the most recent message is done on purpose. It is documented now, I also documented the way MIM finds the message to rectract.

**Comment by chrzaszcz** [comment]
> Thanks for implementing the server support for message retraction! This PR adds this functionality for the RDBMS backends and other backends can be extended later.
> 
> A corner case just came to my mind, which as far as I can see we are handling in a good enough way.
> A user sends two different messages with the same `origin-id`, according to current schema they will be stored. When retracting, only the newer message will be retracted.
> @chrzaszcz do you think it'd be worth to document this behaviour and advise the client to use ids as unique as possible, like `UUIDs`?
> A corner case just came to my mind, which as far as I can see we are handling in a good enough way.
> A user sends two different messages with the same `origin-id`, according to current schema they will be stored. When retracting, only the newer message will be retracted.
> @chrzaszcz do you think it'd be worth to document this behaviour and advise the client to use ids as unique as possible, like `UUIDs`?

Retracting only the most recent message is done on purpose. It is documented now, I also documented the way MIM finds the message to rectract.

**Comment by mongoose-im** [comment]
[8239.1](https://travis-ci.org/esl/MongooseIM/jobs/686457560) / Erlang 22.0 / small_tests / a2636f494ed82dc19c20f481229124fe01d3d613
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2739/8239/small_tests.22.0) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2739/8239/small_tests.22.0/small/ct_run.mongooseim@localhost.2020-05-13_07.33.54/lib.mongooseim.logs/run.2020-05-13_07.33.54/suite.log.html)
